### ISSUE TITLE
How to define a custom operator for conversion from scaled_integer to float

### ISSUE THREAD

**Issue by emiliopaolini** [johnmcfarlane/cnl#848]
Hi all, i have the following problem: i would like to define a custom operator for conversion from the scaled integer type to float.
I think this type of conversion is not defined since when i try to run this snippet:
`#include <cnl/all.h>
#include <cnl/auxiliary/boost.multiprecision.h>
using namespace cnl;

template <int fractionalBits=20>
using X = cnl::scaled_integer<int,cnl::power<-fractionalBits>>;

typedef X<16> float_x;

int main(){
  float_x a =5.721;
  float c = a;
}`
i get the following error: 
error: cannot convert ‘float_x {aka cnl::scaled_integer<int, cnl::power<-16, 2> >}’ to ‘float’ in initialization

Can anyone help me? Thanks!

**Comment by johnmcfarlane** [comment]
Hi @emiliopaolini. Markdown corrupted your code example. Here it is in its entirety:

```c++
#include <cnl/all.h>
#include <cnl/auxiliary/boost.multiprecision.h>
using namespace cnl;

template <int fractionalBits=20>
using X = cnl::scaled_integer<int,cnl::power<-fractionalBits>>;

typedef X<16> float_x;

int main(){
  float_x a =5.721;
  float c = a;
}
```

There is a conversion to `float` but it is marked as `explicit` [here](https://github.com/johnmcfarlane/cnl/blob/v1.1.7/include/cnl/_impl/number/definition.h#L67). You can invoke it by casting `c` to `float` or constructing a `float` from `c` like [this](https://godbolt.org/z/69TrfffnY).

I plan to make use of `explicit` more consistent in #93 but for now, you need to be clear that you want the cast to happen.

# Brief aside about precision

More generally, be aware of precision loss when casting between types that have fractional digits. For example

* `5.721` cannot be represented in binary so `a` is actually initialised with the value, `5.721000000000000085265128291212022304534912109375`,
* neither of those numbers can be represented by `float_x` so `a` is further rounded to `5.7209930419921875`, and
* thus `c` also has value, `5.7209930419921875`

This is still accurate to several decimal places and may well be adequate for your purposes. But it's worth bearing in mind.

I aim to allow lossless initialisation of fixed-point numbers #553 so that

```c++
using namespace cnl::literals;
auto a{5.721_cnl};
```

will produce decimal fixed-point value with no loss of precision. In the meantime, you would need to achieve the same effect using decimal scaled_integer types and division [like so](https://godbolt.org/z/EeGThhfjG):

```c++
  auto a = cnl::scaled_integer<int, cnl::power<-3, 10>>{5721} / 1000;
  std::cout << a << '\n';
  fmt::print("{}\n", a);
  auto c = static_cast<double>(a);
  fmt::print("{:.50}\n", c);
```

This makes `a` precise and increases precision of result of `c` to `5.721000194549560546875`

**Comment by emiliopaolini** [comment]
Thank you very much!
